### ISSUE TITLE
services: Extend service details pages with Firewall information

### ISSUE THREAD

**Issue by garrett** [cockpit-project/cockpit#14321]
In several usability tests, people have looked at (and filtered) the Services page to find a service to open a port.

Example: "Apache isn't working. Oh, the port is blocked." Goes to Services, finds httpd details page... and is sometimes lost at what to do next.

Possible solution: Show port status Service details.

Something like:
- "Port 443/TCP is open."
- "Port 445/TCP is blocked by the firewall."

And then have a button to block or unblock the port, and a link for viewing all firewall rules (which would then go to the Firewall page).

**Comment by marusak** [comment]
Is the idea here that we would just check if service === `httpd.service` and then show this firewall info?
As there are many web servers implementations.

**Comment by garrett** [comment]
The idea is to _somehow_ know the exposed port(s) for a service, for Cockpit to check and show the status on the service details page, and then allow the user to act upon it somehow.

So, for example: they're at Caddy's service page (which is an HTTP server, but specifically not Apache) and see that it's running but the port is blocked. And then they could unblock it (probably selecting a zone, etc.)... and that would affect the firewall rules.

While you and I know that someone should go to the Network page, then Firewall, then add a rule to allow the port... an overwhelming amount of people did not do it that way, hence this issue asking for some integration of Firewall functionality in the Service details page.

**Comment by KKoukiou** [comment]
For socket activated services we can get this information for ListenStream, in the socket unit file. 
And then we can check if the firewall is running, and if that port is blocked or not. 

So for every service, we can check if it has a socket file as well. If so we can get the port from there and continue with the firewall check. 
It looks doable just too much logic maybe for the services page.

**Comment by KKoukiou** [comment]
The part of exposingthe port a service listens to in the service details was straightforward to implement.
The result is now:

![Screen Shot 2021-04-17 at 17 37 29](https://user-images.githubusercontent.com/14921356/115049090-a009fb80-9eda-11eb-8bf8-a6aec19da9a5.png)

The firewall - service page integration however looks not really possible, in a clean way at least.

I faced the following problems while I was doing some effort to move forward:

* Services IDs in the firewall zones service list are not an exact match with the actual service ids, see ssh in firewall API vs sshd.{service, socket} systemd units
* It's hard to know if a port/service is blocked by firewalld, in firewall page we use the oppossite logic, that means, we add services to zones in order to allow traffic for some service, not block it
* There are really much more firewalls on linux apart from firewalld, iptables, nftables, and lastly systemd itself (see http://0pointer.net/blog/ip-accounting-and-access-lists-with-systemd.html) can be used to do filtering of traffic. This does not meant that a hint that firewalld is blocking some specific traffix is bad, just a reminder that it's far from being the only way to do that.

These are just some notes, all in all I think this does not worth to deal with more from an implementation perspective. 
In the new UI one can see the port a service's socket is listening to, so it will probably help the users get a clue, that that specific port needs to be unblocked.

I suggest closing this issue, although this was not addressed completely.

**Comment by martinpitt** [comment]
Thanks for your investigations @KKoukiou ! I already suspected as much. That in fact has been one of the major points of criticism of "classic" firewalls from Lennart & Co - these are completely logically detached from services, so there is no clean relationship between them. Only the `IPAddress{Deny,Allow}=` bits in a unit make sense, but unfortunately that hasn't really taken off at least in Fedora/RHEL so far.

But with all these other firewalls around, you can't really machine-read on the machine itself if a port is accessible from "outside" (which is already fuzzy -- from *which* outside exactly?).

Maybe adding a link to the firewall page next to the "Listen" port list may at least give a hint and some convenience?

**Comment by KKoukiou** [comment]
@martinpitt totally agree with all the above, some points:

* We can easily display IPAddress{Allow, Deny} if these are set, this is straightforward indeed. So I would have done that, but I feel showing these, is like encouraging people to actually use them.
So someone who is not very familiar with services hardening can easily get directed by cockpit UI to do it 'this' way.

*  Adding a link to the firewall page every time there is a 'Listen' row feels too much. Most of the times there will not actually be a problem with blocked traffic and the users will just get extra information

**Comment by martinpitt** [comment]
Good point, yes. So indeed I agree that displaying IPAddress* as fields makes total sense! Maybe it nudges people a little bit more towards doing it the "modern" way :grin:

**Comment by marusak** [comment]
So it seems we agree that we should not show what this PR proposes - to show if port is open or not.
Two separate ideas that came up here:
- Add link to firewall next to `Listen`. While I can see benefit of this, we would need to communicate it properly so it is clear, that the firewall might not have anything to do with it. Maybe (?) with tooltip like "Ports might be blocked by firewall. Go to firewall".
- Add `IPAddress{Deny,Allow}` : we can add this by adding separate proxy to `org.freedesktop.systemd1.Service` interface. And then to `Socket`, `Mount`... I wonder if it is worth doing that :man_shrugging: Also the output is of this is somewhat ugly: (It is array of two items, first it IPv4 the other IPv6, one represent 127.0.0.0/8 the other ::1/128 - but in the file you could write this down as `localhost`)
```
$ busctl get-property org.freedesktop.systemd1 /org/freedesktop/systemd1/unit/chrony_2dwait_2eservice org.freedesktop.systemd1.Service IPAddressAllow
a(iayu) 2 2 4 127 0 0 0 8 10 16 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 128
```
or as JS sees it in base64
```
[[2,"fwAAAA==",8],[10,"AAAAAAAAAAAAAAAAAAAAAQ==",128]]
```

So it would need some parsing and possibly also being able to convert it to known values as `any`, `localhost` - which you can use in the definition files.
